Speed up deposit and transactions view loads

stejbac opened the pull request:
<!-- 
- make yourself familiar with the CONTRIBUTING.md if you have not already (https://github.com/bisq-network/bisq/blob/master/CONTRIBUTING.md)
- make sure you follow our [coding style guidelines][https://github.com/bisq-network/style/issues)
- pick a descriptive title
- provide some meaningful PR description below
- create the PR
- in case you receive a "Change request" and/or a NACK, please react within 30 days. If not, we will close your PR and it can not be up for compensation.
- After addressing the change request, __please re-request a review!__ Otherwise we might miss your PR as we tend to only look at pull requests tagged with a "review required".
-->

This PR speeds up the loading of the deposit ("Receive funds") and transactions views (under the "Funds" tab), by adding some caches to `WalletService` and `DisputeListService`, lazily loading tx confidence tooltips for the two respective lists of items (`DepositListItem` and `TransactionsListItem`) and applying some low level optimisations to `TransactionsAwareTrade.isRelatedToTransaction`.

This also fixes a fairly minor bug in the transactions view due to a broken `TranscationsAwareTrade.isRefundPayoutTx` method, as well as a somewhat serious memory leak in the deposit view due to missing `BalanceListener` removals from the `BtcWalletService` instance.

The performance issues in these two views had started to become severe with the number of transactions and addresses now in my Bisq wallet. The deposit view was the worst, taking around 25 seconds to load, with bottlenecks in `WalletService.getTransactionConfidence` & `WalletService.getNumTxOutputsForAddress` as seen with JProfiler:

![Screenshot from 2021-01-27 02-41-06](https://user-images.githubusercontent.com/54855381/105935157-621ed900-6049-11eb-91bb-2cf83e28984e.png)

The transactions view (with 1210 rows) was taking around 5 seconds to load, with bottlenecks in `TransactionsAwareTrade.isRelatedToTransaction` & `Tooltip.<init>`:

![Screenshot from 2021-01-27 02-41-25](https://user-images.githubusercontent.com/54855381/105935191-6f3bc800-6049-11eb-8d6c-526ba9a96aa8.png)

They both seem significantly faster (in my Bisq client) with the changes in this PR, with the biggest improvement to the deposit view via the two caches added to `WalletService`. The quadratic time issue (# transactions * # past trades) causing the slowness of the transactions view is alleviated but not completely fixed by these changes. It may be possible to do more extensive lazy loading of the relevant `TransactionsListItem` fields to get further speedups.

stejbac committed:
Refactor: add & remove wallet listeners from the same class

Make the WalletService.walletEventListener field private and add it via
a protected method defined in the base class, addListenersToWallet(), so
that the setup code in the two subclasses (Bsq|Btc)WalletService can be
deduplicated and more easily kept in sync with the listener removal code
in WalletService.shutDown().

Also remove some unnecessary deprecation warning suppressions.

stejbac committed:
Use cache for WalletService.getNumTxOutputsForAddress

Use a guava Multiset to cache the total number of tx outputs (out of the
live txs in the user's wallet) with a given address. Since this requires
a scan of the entire tx set, compute all the counts in one go and store
in an ImmutableMultiset<Address>. Invalidate the entire cache any time a
tx set change occurs, by attaching a WalletChangeEventListener to the
wallet (using a direct executor for immediate effect).

This is to fix a quadratic time bug in DepositView, which uses the count
to determine if a given address in the BTC wallet is used/unused.

stejbac committed:
Use cache for WalletService.getConfidenceForAddress

Use a guava SetMultimap (a many-to-many mapping without duplicates) to
cache the set of live txs in the user's wallet with a given address as
an input or output. As with the cache of output counts from the previous
commit, compute all the tx sets in one go (by a tx stream followed by a
map inverse) and store in an ImmutableSetMultimap<Address, Transaction>,
invalidating the entire cache immediately upon each wallet change event.

This is to fix another (larger) quadratic time bug in DepositView, when
getting the confidence (i.e. confirmation count) of each wallet address.

Also simplify getTransactionConfidence & onTransactionConfidenceChanged
methods slightly, which generated (possibly unintentionally) repeating &
singleton lists of TransactionConfidence objects to pass to
WalletService.getMostRecentConfidence(..) respectively.

stejbac committed:
Optimise TransactionAwareTrade.isRelatedToTransaction

Attempt to remove a bottleneck during the transactions view load, as
revealed by JProfiler, by optimising the code to determine if any given
transaction and trade are related. Since both these sets will tend to
grow linearly with time, causing quadratic slowdown of TransactionsView,
try to alleviate (without completely fixing) the problem.

To do this, add a cached set of disputed trade IDs to DisputeListService
so that TransactionAwareTradable.is(Dispute|RefundPayout)Tx can be made
O(1) in the common case that the given trade is not involved in any
dispute. Also avoid calling Sha256Hash::toString by passing tx IDs as
Sha256Hash objects directly to is(Deposit|Payout)Tx, and short circuit
an expensive call BtcWalletService.getTransaction in isDelayedPayoutTx,
in the common case, by pre-checking the transaction locktime.

This also fixes a bug in isRefundPayoutTx whereby it incorrectly returns
false if there are >1 disputes in the list returned by RefundManager but
the given trade is not involved in the last one.

stejbac committed:
Add missing WalletService.removeBalanceListener to fix leak

Fix a serious memory leak in DepositListItem due to missing removal of
the BalanceListener (one per item) from BtcWalletService. This prevented
GC of the entire list of items, which was observed to leak ~500 MB in
JProfiler after repeated switching (several dozen times) between tabs.

stejbac committed:
Load (Deposit|Transactions)ListItem tooltips lazily

Add a nested class of lazy fields to (Deposit|Transactions)ListItem,
together with an associated memoised Supplier, lazyFieldsSupplier, which
initialises them all the first time any one of them is requested. Move
the txConfidenceIndicator & tooltip fields to this class, so that they
are only loaded when the given address/tx row is displayed in the
deposit/transactions views, respectively.

This prevents a minor bottleneck, as profiling indicates that creating a
tooltip for each tx confidence indicator in the list is quite expensive
and takes up around half the rendering time. (There may be 1000's of txs
or addresses in an old wallet.)

chimp1984 commented on core/src/main/java/bisq/core/btc/wallet/WalletService.java:
```diff
@@ -408,18 +430,15 @@ public TransactionConfidence getConfidenceForTxId(String txId) {
-                .map(o -> tx.getConfidence())
-                .collect(Collectors.toList());
-        return getMostRecentConfidence(transactionConfidenceList);
+    private TransactionConfidence getTransactionConfidence(Transaction tx, Address address) {
```
Can we add a @Nullable annotation (was before as well nullable as getMostRecentConfidence is nullable).

stejbac replied:
OK, I agree it probably should be `@Nullable`.

chimp1984 commented on core/src/main/java/bisq/core/btc/wallet/WalletService.java:
```diff
@@ -408,18 +430,15 @@ public TransactionConfidence getConfidenceForTxId(String txId) {
+        boolean matchesAddress = getOutputsWithConnectedOutputs(tx).stream()
+                .anyMatch(output -> address != null && address.equals(getAddressFromOutput(output)));
+
+        return matchesAddress ? getMostRecentConfidence(List.of(tx.getConfidence())) : null;
```
I think we have an issue here:
Assuming there was the same address used as connected output of an input and as the output of the given tx.
We had then 2 different txs in the previous version and the getMostRecentConfidence took the most recent one (the outputs as the connected output of the input could not be newer). Now we drop any of the 2 txs by `anyMatch` and use only the one of the tx (tx output). This would be correct in case we had 2 addresses connected to that tx.
But if the address was only used in a connected output of the input we would take the wrong txConfidence as we would take the actual tx and not the tx associated with the connected output of the input.

stejbac replied:
Yes, I wasn't completely sure of the intent of the original version and tried to preserve its behaviour, but it looks like it may have been broken. The original stream pipeline had the line:

> .map(o -> tx.getConfidence())

But `tx` is the underlying method parameter and so this just fed a repeating list of the same `TransactionConfidence` object into `getMostRecentConfidence`.

I think it may be fixed by restoring the original pipeline, but using the lines:

> .flatMap(o -> Stream.ofNullable(o.getParentTransaction()))
> .map(Transaction::getConfidence)

in place of the `o -> tx.getConfidence` mapping.

This would make the method return null if (there was no matching output and) none of the tx inputs with the given address has a connecting parent tx (though I'm not sure if the latter ever happens). I guess that would be the right behaviour.

chimp1984 replied:
Oh, yes looks like that `.map(o -> tx.getConfidence())` was a bug. I think intention was to find all txs where the address was used for receiving or sending BTC. And take the most recent confidence. Unforunately the design flaw to use addresses so dominantely causes many issues in many area. Maybe we can slowely try to convert the domain model to use txos as the core entity and if the UI needs to show addresses we grab the collection of txos related to that address.

chimp1984 reviewed (changes requested):
I think the getTransactionConfidence method is not correct. Otherwise all looks good. Was not aware of the guava Multimap framework...Could be used in many places...

stejbac committed:
Fix broken WalletService.getTransactionConfidence(tx, address)

Undo the earlier simplification changes to getTransactionConfidence,
which preserved its original but broken behaviour. Fix the original
stream pipeline so that each matching tx input maps to the confidence of
the connected parent tx (if any), not the child tx. In this way, it
correctly considers parent tx confidences when determining the most
recent confidence of all the matching inputs & outputs.

Before it was simply feeding a repeating list of identical objects into
getMostRecentConfidence, via the erroneous line:

  .map(o -> tx.getConfidence())

(Also add a missing @Nullable annotation & make getMostRecentConfidence
private instead of protected.)

Finally, simplify BisqWalletListener.onTransactionConfidenceChanged, by
no longer feeding a singleton list into getMostRecentConfidence whose
element is already a return value of that method, as that is a no-op.

chimp1984 reviewed (approved):
utACK
